### ISSUE TITLE
Difference between DGS and naive trilinear interp?

### ISSUE THREAD

**Issue by fishfishson** [zhusz/ICLR22-DGS#1]
Hi zhu, thx for your great research work!
I read your paper but was still confused with your DGS algorithm. Does it mean a differentiable tri/bilinear interp? If so, what's the difference between DGS and naive trilinear interp like

1. index 8 corner features
2. calculate 8 weights for query point (x,y,z) 
3. get final features by weighted-sum of above corner featrues.

The above method is also differentiable wrt query point and corner features, I think.
Looking forward to your reply!

**Comment by zhusz** [comment]
Hi thanks for your interest in our work! 

Our spatial gradient during the forward propagation is the same as the bilinear interpolation (as the three steps listed above). Our main motivation is that the naive implementation will result in this [error](https://github.com/zhusz/ICLR22-DGS/blob/master/src/P/Pscannet/D0gridsamplecannotwork/extension.py#L87), and in particular, it will prompt up during the backward process at the sampling operation. We thus devise our cuda implementation for tracking the spatial gradient during the forward process ([here](https://github.com/zhusz/ICLR22-DGS/blob/master/src/P/Pscannet/trainer.py#L985)). We found training using per-query-point occupancy label of the real-scan data (without using the gradient loss) seems to be somewhat difficult engineeringly, and not generalize well to external unseen test scenes (e.g. the [demo videos](https://zhusz.github.io/ICLR22-DGS-Webpage/)).

We later following one of the reviewers' hint to obtain a way to achieve closed-form gradient training without cuda implementation via [this](https://github.com/zhusz/pp/blob/master/src/versions/codes_py/toolbox_3D/pydgs_v1.py), where one can use these functions to replace `torch.nn.functional.grid_sample`. No error will prompt up here. But it can only be applied to one-dimensional output scenario (e.g. occupancy is OK, rgb / semantics is not OK). There is also some other scenarios where this has limitation for use.

Hope this help.
